system/jlib/jsocket.cpp

```diff
@@ -1921,257 +1921,149 @@ int CSocket::wait_write(unsigned timeout)
-        if (elapsed<timeoutms)
-            timeleft = timeoutms-elapsed;
-        else
-            timeleft = 0;
+    CCycleTimer timer;
+    while (true)
+    {
+        unsigned remainingMs = timer.remainingMs(timeoutMs);
+        int rc = wait_read(remainingMs);
+        if (rc < 0)
+            THROWJSOCKTARGETEXCEPTION(SOCKETERRNO());
+        else if (rc == 0)
+            THROWJSOCKTARGETEXCEPTION(JSOCKERR_timeout_expired);
+
         unsigned retrycount=100;
 EintrRetry:
-        if (sockmode==sm_udp_server) { // udp server        
+        if (sockmode==sm_udp_server) // udp server
+        {
             DEFINE_SOCKADDR(u);
             socklen_t ul=sizeof(u);
-            rc = recvfrom(sock, (char*)buf + size_read, max_size - size_read, 0, &u.sa,&ul);
+            rc = recvfrom(sock, (char*)buf + sizeRead, max_size - sizeRead, 0, &u.sa,&ul);
             getSockAddrEndpoint(u,ul,returnep);
         }
-        else {
-            rc = recv(sock, (char*)buf + size_read, max_size - size_read, 0);
+        else
+            rc = recv(sock, (char*)buf + sizeRead, max_size - sizeRead, 0);
+
+        if (rc > 0)
+        {
+            sizeRead += rc;
+            if (sizeRead >= min_size)
+                break;
         }
-        if (rc < 0) {
-            int err = SOCKETERRNO();
-            if (BADSOCKERR(err)) {
-                // don't think this should happen but convert to same as shutdown while investigation
-                LOGERR2(err,1,"Socket closed during read");
-                rc = 0;
-            }
-            else if ((err==JSE_INTR)&&(retrycount--!=0)) {
-                if (sock==INVALID_SOCKET)
-                    rc = 0;         // convert an EINTR after closed to a graceful close
+        else
+        {
+            if (rc < 0)
+            {
+                int err = SOCKETERRNO();
+                if (BADSOCKERR(err))
+                {
+                    // don't think this should happen but convert to same as shutdown while investigation
+                    LOGERR2(err,1,"Socket closed during read");
+                    rc = 0;
+                }
+                else if ((err==JSE_INTR)&&(retrycount--!=0))
+                {
+                    if (sock==INVALID_SOCKET)
+                        rc = 0;         // convert an EINTR after closed to a graceful close
+                    else
+                    {
+                        LOGERR2(err,1,"EINTR retrying");
+                        goto EintrRetry;
+                    }
+                }
                 else
                 {
-                    LOGERR2(err,1,"EINTR retrying");
-                    goto EintrRetry;
+                    if (nonblocking && (err == JSE_WOULDBLOCK || err == EAGAIN)) // if EGAIN or EWOULDBLOCK - no more data to read
+                    {
+                        if (0 == min_size) // if min_read is 0, then whatever we have read so far is good enough (even if 0)
```

Comment:
>if (sizeRead >= min_size)? 

as with the similar code in secure socket (and unlike the same test when rc==0), I think this is okay as it is, because if min_size > 0, and sizeRead > 0, then it would have already checked sizeRead >= min_size in the (rc > 0) block.

Reply:
ok.  Comment could be improved then to avoid confusion - since this will never have read anything so far.

Reply:
> ok. Comment could be improved then to avoid confusion - since this will never have read anything so far.

changed comments, and squashed.
@ghalliday - please review/merge.